### PR TITLE
Introduce CenterCropPad, add `axes` to Pad

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

**Description**
- Adds a new function: CenterCropPad-18.
- CenterCropPad is a typical operation in image processing application. 
- It has been implemented as a function in terms of Pad and Slice
- Pad-18 to have axes attribute

**Motivation and Context**
- Why is this change required? What problem does it solve?
Enables writing image preprocessing pipelines (such as ResNet-50)

### CODE REVIEW

File: onnx/defs/tensor/defs.cc [hunk 1]
```diff
@@ -3912,4 +3990,156 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+          }
+          int j = 0;
+          for (int axis : axes) {
+            out_dims[axis]->set_dim_value(shape[j++]);
+          }
+        })
+        .SetContextDependentFunctionBodyBuilder(
+            [](const FunctionBodyBuildContext& ctx, const OpSchema& schema, FunctionProto& functionProto) {
+              const auto& input_shape = ctx.getInputType(0)->tensor_type().shape();
+              const int64_t input_rank = input_shape.dim_size();
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc:4109]: Unfortunately, this is going to be a problem. The shape we can get is only a static-shape, not a dynamic-shape. Statically, we may not know the rank of the input.

Comment (gramalingam) [reply onnx/defs/tensor/defs.cc:4109]: Instead, one alternative is to get the (full) shape, convert the axes attribute into a tensor, and use it to index into the full-shape using some form of gather op.

File: onnx/defs/tensor/defs.cc [hunk 2]
```diff
@@ -3646,37 +3610,46 @@ ONNX_OPERATOR_SET_SCHEMA(
...
 
               auto axes_attr = ctx.getAttribute("axes");
               std::vector<int64_t> axes;
-              if (axes_attr) {
+              if (axes_attr) {  // axes provided, need to work on a subset of dimensions
                 axes = RetrieveValues<int64_t>(*axes_attr);
-              } else {
-                axes.resize(input_rank);
-                std::iota(axes.begin(), axes.end(), 0);
+                builder.Const("axes_input", axes);
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc:3620]: I don't see `axes` being initialized above. Either it needs to be defined, or a better option is to use
```cpp
    builder.Add("axes_input = Constant <value_ints : ints = @axes>()");
```

Comment (jantonguirao) [reply onnx/defs/tensor/defs.cc:3620]: I don't understand this comment. `axes` is declared in line 3617 and the values are read in line 3619.
That being said, your suggestion seems better. I'll change it.

Comment (jantonguirao) [reply onnx/defs/tensor/defs.cc]: Done

**Comment by gramalingam** [comment]
LGTM, thanks. Have the test-cases been validated with ORT? Unfortunately, there is no other way to validate function definitions.

**Comment by jantonguirao** [comment]
> LGTM, thanks. Have the test-cases been validated with ORT? Unfortunately, there is no other way to validate function definitions.

Yes. I have validated the test cases that don't depend on `Pad-18` having support for `axes`

**Review by p-wysocki — COMMENTED** [review]
Left one comment, LGTM otherwise.

File: onnx/defs/tensor/defs.cc [hunk 3]
```diff
@@ -3315,11 +3316,24 @@ ONNX_OPERATOR_SET_SCHEMA(
             true,
             1,
             OpSchema::NonDifferentiable)
+        .Input(
+            3,
+            "axes",
+            "1-D tensor of axes that `pads` apply to. Negative value means counting dimensions "
+            "from the back. Accepted range is [-r, r-1] where r = rank(data). Behavior is undefined if an "
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply onnx/defs/tensor/defs.cc:3323]: Done

**Review by gramalingam — APPROVED** [review]

**Comment by fdwr** [comment]
- What is the behavior for odd input sizes? Floor? Ceil?
- When padded, what is the fill value? Zero? Edge repetition?

The documentation currently lacks these salient details: https://onnx.ai/onnx/operators/onnx__CenterCropPad.html#l-onnx-doc-centercroppad

**Comment by jantonguirao** [comment]
> * What is the behavior for odd input sizes? Floor? Ceil?
> * When padded, what is the fill value? Zero? Edge repetition?
> 
> The documentation currently lacks these salient details - https://onnx.ai/onnx/operators/onnx__CenterCropPad.html#l-onnx-doc-centercroppad. Is the following correct?
> 
> > If the input dimensions are bigger than the crop shape, a centered cropping window is extracted from the input. If the input dimensions are smaller than the crop shape, the input is padded on each side equally <**with zero values**>, so that the input is centered in the output. <**If the input dimensions are odd relative to the cropping window, the offset is floored**>.

I've created a PR to enhance the documentation:
https://github.com/onnx/onnx/pull/6464

**Pull request merged**